tensorflow/lite/micro/micro_profiler.h

```diff
@@ -33,67 +46,185 @@ class MicroProfiler : public MicroProfilerInterface {
+          MicroPrintf("%d,%s,%" PRIu64 ".%" PRIu64 ",%" PRIu32, i, tags_[i],
+                      us / 1000, us % 1000, ticks);
+#endif
+        }
+        break;
+    }
+#endif  // !defined(TF_LITE_STRIP_ERROR_STRINGS)
+  }
 
-  // Prints  total ticks for each unique tag in CSV format.
-  // Output will have one row for each unique tag along with the
-  // total ticks summed across all events with that particular tag.
-  void LogTicksPerTagCsv();
+  // Prints the profiling information of each of the events in human readable
+  // form, grouped per tag, sorted by execution time.
+  void LogGrouped(MicroProfilerLogFormat format) {
+#if !defined(TF_LITE_STRIP_ERROR_STRINGS)
+    for (int i = 0; i < num_events_; ++i) {
+      // Find if the tag already exists in uniqueTags
+      TagGroup& tag_group = GetTagGroup(tags_[i]);
+
+      uint32_t ticks = end_ticks_[i] - start_ticks_[i];
+      tag_group.tag = tags_[i];
+      tag_group.ticks += ticks;
+      tag_group.tag_count++;
+    }
+
+    SortTagGroups();
+
+    switch (format) {
+      case MicroProfilerLogFormat::HumanReadable: {
+        MicroPrintf("Cumulative event times:");
+        MicroPrintf("%-8s %-32s %-12s %-12s", "Count", "Tag", "Ticks", "Time");
+        uint64_t total_ticks = 0;
+        uint64_t us;
+        for (int i = 0; i < num_tag_groups_; ++i) {
+          total_ticks += tag_groups_[i].ticks;
+          us = TicksToUs(tag_groups_[i].ticks);
+          MicroPrintf("%-8d %-32s %-12d %" PRIu64 ".%03" PRIu64 " ms",
+                      tag_groups_[i].tag_count, tag_groups_[i].tag,
+                      tag_groups_[i].ticks, us / 1000, us % 1000);
+        }
+        us = TicksToUs(total_ticks);
+        MicroPrintf("\nTotal time: %" PRIu64 ".%03" PRIu64 " ms (%lld ticks)",
+                    us / 1000, us % 1000, total_ticks);
+        break;
+      }
+      case MicroProfilerLogFormat::Csv: {
+        MicroPrintf("\"Tag\",\"Total ms\",\"Total ticks\"");
+        for (int i = 0; i < num_tag_groups_; ++i) {
+          uint64_t us = TicksToUs(tag_groups_[i].ticks);
+          MicroPrintf("%s, %" PRIu64 ".%" PRIu64 ", %u", tag_groups_[i].tag,
+                      us / 1000, us % 1000, tag_groups_[i].ticks);
+        }
+        break;
+      }
+    }
+#endif  // !defined(TF_LITE_STRIP_ERROR_STRINGS)
+  }
 
  private:
-  // Maximum number of events that this class can keep track of. The
-  // MicroProfiler will abort if AddEvent is called more than kMaxEvents number
-  // of times. Increase this number if you need more events.
-  static constexpr int kMaxEvents = 4096;
-
-  const char* tags_[kMaxEvents];
-  uint32_t start_ticks_[kMaxEvents];
-  uint32_t end_ticks_[kMaxEvents];
+  const char* tags_[MAX_EVENTS];
+  uint32_t start_ticks_[MAX_EVENTS];
+  uint32_t end_ticks_[MAX_EVENTS];
   int num_events_ = 0;
+  int num_tag_groups_ = 0;
 
-  struct TicksPerTag {
+  struct TagGroup {
     const char* tag;
     uint32_t ticks;
+    uint32_t tag_count;
```

Comment:
Only used by the "human readable" log format.

Reply:
What should I do about that? #1290 introduced

```
struct TicksPerTag {
    const char* tag;
    uint32_t ticks;
  };
  // In practice, the number of tags will be much lower than the number of
  // events. But it is theoretically possible that each event to be unique and
  // hence we allow total_ticks_per_tag to have kMaxEvents entries.
  TicksPerTag total_ticks_per_tag[kMaxEvents];
```

which was "only" used by the CSV log format.

Reply:
@andresovela 

The comment is really just there for @suleshahid to see.